### DISCUSSION TITLE
Proper way to use Allan Variance

### DISCUSSION THREAD

**Discussion by BrunoOtavio98** [rodralez/NaveGo#53] [Q&A] [answered]
Hi, im trying to use your Kalman Filter architecture in my navegation system and i am new in this field. How should i proceed to get the accelerometer and gyroscope data to pass to allan_imu and get the errors parameters? Now i tried to leave them stationary for 20 minutes [MatLab Ref,](https://www.mathworks.com/help/nav/ug/inertial-sensor-noise-analysis-using-allan-variance.html) thats correct? Or they should be in movement? 

If a run in this way, i get this error:

error: dtime(1): out of bound 0 (dimensions are 1x0)
error: called from
    allan_overlap at line 534 column 5
    allan_imu at line 181 column 30

Also in allan_overlap i get this two lines:

allan_overlap: regular data (0 freq data points @ 130 Hz)
allan_overlap: phase data generated from fractional frequency data (N=1).

**Comment by ravindraji4** [comment, marked as answer]
Hi Bruno,

Firstly, the data should be collected with IMU at rest and this should be inputted to calculate the different Allan Variance Values. 

Secondly, I tried the inbuilt Matlab example and it works. 

It would be easier. If, Could you please upload the data. So that we can find the problem. 
Just writing the error/encountered problem, it is hard to say/explain the potential solution! 

Best regards, 
Ravindra

**Reply by BrunoOtavio98** [reply]
Hi @ravindraji4 , thank you for answer me,  so, after i set verbose with 0, the allan variance have work. I left the imu at rest for 2 hours. But the result of the sensor fusion was horrible, in fact worst than the other model that i have made, thats is, in fact, more  simple than that. Since im using a low cost imu (BNO085) and a low cost GNSS (Quectel L76) im not expecting something extremely precise. But the model with only position and velocity gave me best results, at this model, i used the R-P-Y angles that the imu itself provides me to compute the velocity and position at NED frame. I cant find where i made a mistake running the sensor fusion of NaveGO. The first image was a displacement with a model that i have made, the second is one with NaveGo. In blue the kalman output and in red the GPS only output from both models.

![xsWka](https://user-images.githubusercontent.com/40346647/109394388-864e2e00-7905-11eb-9a56-7581943dbf9d.jpg)
![Capturar](https://user-images.githubusercontent.com/40346647/109394511-299f4300-7906-11eb-9605-73dda330a70e.PNG)

Here is a link of the dataset that i used
[GPSdata.txt](https://github.com/rodralez/NaveGo/files/6054973/GPSdata.txt)
[IMUBNO085data.txt](https://github.com/rodralez/NaveGo/files/6054974/IMUBNO085data.txt)

And the link of the IMU model for the allan variance.
https://drive.google.com/file/d/1az16b08X8RDZPk0jLAvtorNt1wdfZXoL/view?usp=sharing

For the imu dataset, the data is: accelx, accely, accelz, gyrox, gyroy, gyroz, magx, magy, magz, roll, pitch, yaw, timestamp. I put the R-P-Y on the dataset only for comparison with the NaveGo output. The y axis is alligned with the vehicle acceleration and the x axis with lateral acceleration. The gyro is already at rad/s.

For the GPS dataset: Lat, lon, alt, Vel, Vel_N, Vel_E, Vel_D, and timestamp. Since my gps dont output velocity in NED frame i used the heading of the imu to get this information, and the down velocity i got with the difference between two altitudes and their timestamps.

**Reply by ravindraji4** [reply]
Hi Bruno!
Thank you very much for uploading the datasets. 
The plotted the GPSdata from the second track. 
![image_2021-03-08_101236](https://user-images.githubusercontent.com/26116885/110300116-cd6bab80-7ff6-11eb-851d-7e02aa2f3d4f.png)




I compared the GPS and IMU data and found that the timestamps are not matching. therefore, we can't relate the recorded data from both datasets. 
For example....the GPS time stamp starts with 88.12, 88.12+0.1099999 which should increment. The same is valid with IMU 41.47, 41.47+0.009.

![Github1](https://user-images.githubusercontent.com/26116885/110304490-f5114280-7ffb-11eb-8521-9d482e2e600e.PNG)



Best is if they match with best possible decimal. In other case, they could roughly match!
Then you can try again. Or if you wish to upload the data.

**Reply by BrunoOtavio98** [reply]
Hi @ravindraji4  yes, thats my fault, thats because im not logging gps data untill occurs a valid position, so imu have started to log before gps, but, since both are running on the same MCU the gps timestamp = 88.12 its equal of imu timestamp = 88.12, i will fix this and re-upload the data.

**Reply by rodralez** [reply]
> For the imu dataset, the data is: accelx, accely, accelz, gyrox, gyroy, gyroz, magx, magy, magz, roll, pitch, yaw, timestamp. I put the R-P-Y on the dataset only for comparison with the NaveGo output. The y axis is alligned with the vehicle acceleration and the x axis with lateral acceleration. The gyro is already at rad/s.
 
> For the GPS dataset: Lat, lon, alt, Vel, Vel_N, Vel_E, Vel_D, and timestamp. Since my gps dont output velocity in NED frame i used the heading of the imu to get this information, and the down velocity i got with the difference between two altitudes and their timestamps.

Bruno, if your GNSS sensor does not deliver NED velocity, you should try to get these data from the GNSS positioning measurements. You can try to process the GNSS data with RTKPOST from RTKLIB program and extract the $VELACC field. You have to choice "States" in the Output Solution Status. 
<img src="https://user-images.githubusercontent.com/5432060/113207312-d8f17180-9246-11eb-8c01-1c161fd12487.png" alt="drawing" width="400"/>

Then, you can use the function rtkpost_read.m from NaveGo to extract NED velocities.

If  NED velocity is estimated from IMU data, these data will be strongly correlated. This situation should be avoided in order to improve the performance of the Kalman filter.

Regards.

**Reply by rodralez** [reply]
> And the link of the IMU model for the allan variance.
> https://drive.google.com/file/d/1az16b08X8RDZPk0jLAvtorNt1wdfZXoL/view?usp=sharing

Bruno, I checked the static data from your IMU. Accelerometers X and Y seems to have an excessive static bias (-2.1133, 4.7109) which points out that the IMU was not corrected leveled during the recording session. On the other hand, gyros data is mostly zero, which is odd.

Best regards.

**Reply by BrunoOtavio98** [reply]
Hi @rodralez, about accelerometer, i think that's because the imu was inside the development product and  i not sure how the imu is placed inside it. About the gyro, thats strange actually, i will see whats happening inside the firmware.

**Reply by BrunoOtavio98** [reply]
> 
> 
> > For the imu dataset, the data is: accelx, accely, accelz, gyrox, gyroy, gyroz, magx, magy, magz, roll, pitch, yaw, timestamp. I put the R-P-Y on the dataset only for comparison with the NaveGo output. The y axis is alligned with the vehicle acceleration and the x axis with lateral acceleration. The gyro is already at rad/s.
> 
> > For the GPS dataset: Lat, lon, alt, Vel, Vel_N, Vel_E, Vel_D, and timestamp. Since my gps dont output velocity in NED frame i used the heading of the imu to get this information, and the down velocity i got with the difference between two altitudes and their timestamps.
> 
> Bruno, if your GNSS sensor does not deliver NED velocity, you should try to get these data from the GNSS positioning measurements. You can try to process the GNSS data with RTKPOST from RTKLIB program and extract the $VELACC field. You have to choice "States" in the Output Solution Status.
> <img alt="drawing" width="400" src="https://user-images.githubusercontent.com/5432060/113207312-d8f17180-9246-11eb-8c01-1c161fd12487.png">
> 
> Then, you can use the function rtkpost_read.m from NaveGo to extract NED velocities.
> 
> If NED velocity is estimated from IMU data, these data will be strongly correlated. This situation should be avoided in order to improve the performance of the Kalman filter.
> 
> Regards.

But to use RKTLIB, GNSS must provide raw data, right?  I talked with quectel and L76 only deliver final positioning data.

**Reply by BrunoOtavio98** [reply]
Hi, Im re-uploading the dataset, i changed the ned velocity with one thats not dependent of imu heading. Together, one "ground truth" displacement with RTK receiver.
[IMUBNO085data.txt](https://github.com/rodralez/NaveGo/files/6421792/IMUBNO085data.txt)
[GPSdata 2.txt](https://github.com/rodralez/NaveGo/files/6421795/GPSdata.2.txt)
[gpsRef.txt](https://github.com/rodralez/NaveGo/files/6421808/gpsRef.txt)

**Reply by BrunoOtavio98** [reply]
And this is IMU static data to allan variance.
[https://drive.google.com/file/d/1cmiveOUbyou3R2IGo7KDycCSwq3M7sYH/view?usp=sharing](url)

**Reply by rodralez** [reply]
Hi Bruno,

Just give me a couple of days to process your new data.

Best regards.

**Reply by rodralez** [reply]
> Hi, Im re-uploading the dataset, i changed the ned velocity with one thats not dependent of imu heading. Together, one "ground truth" displacement with RTK receiver.
> [IMUBNO085data.txt](https://github.com/rodralez/NaveGo/files/6421792/IMUBNO085data.txt)
> [GPSdata 2.txt](https://github.com/rodralez/NaveGo/files/6421795/GPSdata.2.txt)
> [gpsRef.txt](https://github.com/rodralez/NaveGo/files/6421808/gpsRef.txt)

Hello Bruno,

I attach just a template to help you start processing your IMU and GNSS data by using NaveGo.

I think your data is not correctly sampled, or in other words, the sampling time of both sensors seems to be unstable. Check the output of test_time_quality().

I hope this could help you to use NaveGo!

Best wishes.

[bruno_navego.zip](https://github.com/rodralez/NaveGo/files/6469133/bruno_navego.zip)

**Reply by BrunoOtavio98** [reply]
Hello @rodralez, first of all, thank you for your support. Ok, i will see the code.

**Comment by rodralez** [comment]
> Hi, im trying to use your Kalman Filter architecture in my navegation system and i am new in this field. How should i proceed to get the accelerometer and gyroscope data to pass to allan_imu and get the errors parameters? Now i tried to leave them stationary for 20 minutes [MatLab Ref,](https://www.mathworks.com/help/nav/ug/inertial-sensor-noise-analysis-using-allan-variance.html) thats correct? Or they should be in movement?

Hello Bruno,

It is convenient to have between 2 and 12 hours of static data for an Allan variance analysis.

More info: 

 IEEE-SA Standards Board. IEEE Standard Specification Format
Guide and Test Procedure for Single-Axis Interferometric Fiber Optic
Gyros. ISBN 1-55937-961-8. September 1997.

 Naser El-Sheimy et at. Analysis and Modeling of Inertial Sensors
Using Allan Variance. IEEE TRANSACTIONS ON INSTRUMENTATION AND
 MEASUREMENT, VOL. 57, NO. 1, JANUARY 2008.

 Oliver J. Woodman. An introduction to inertial navigation. Technical
Report. ISSN 1476-2986. University of Cambridge, Computer Laboratory.
August 2007.

Best regards.

**Reply by BrunoOtavio98** [reply]
Thank you for your answer @rodralez  i will read the papers.

**Reply by rodralez** [reply]
By the way, I am reading the article about Allan variance by MathWorks that you provided above is excellent. It  is a concise and clear explanation of the the Allan variance. 

Best regards.
